pinot-core/src/main/java/org/apache/pinot/core/segment/processing/genericrow/FileWriter.java

```diff
@@ -0,0 +1,27 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.core.segment.processing.genericrow;
+
+import java.io.IOException;
+
+
+public interface FileWriter<T> {
```

Comment:
**Suggestion** -  Instead of AdaptiveConstrainsChecker, may be we can introduce an AdaptiveConstraintsWriter interface, that takes in the GenericWriter and delegates the write calls to the GenericWriter. AdaptiveConstraintsWriter can still have the canWrite() api (to terminate mapper phase), in addition to write() api.
Whenever write is involved, bytes are automatically tracked in AdaptiveConstraintsWriter implementation (like AdaptiveSizeBasedWriter) and we dont need to explicitly update bytes or reset counters.
That way we dont need to expose adaptive constraints logic in mapper (like updating bytes, resetting counter etc.). When a new writer is initialized, counters are automatically reset.